plugins/experimental/log_requests/log_requests.c

```diff
@@ -0,0 +1,301 @@
+                                TS_HTTP_STATUS_INTERNAL_SERVER_ERROR,
+                                TS_HTTP_STATUS_NOT_IMPLEMENTED,
+                                TS_HTTP_STATUS_BAD_GATEWAY,
+                                TS_HTTP_STATUS_SERVICE_UNAVAILABLE,
+                                TS_HTTP_STATUS_GATEWAY_TIMEOUT,
+                                TS_HTTP_STATUS_HTTPVER_NOT_SUPPORTED,
+                                TS_HTTP_STATUS_VARIANT_ALSO_NEGOTIATES,
+                                TS_HTTP_STATUS_INSUFFICIENT_STORAGE,
+                                TS_HTTP_STATUS_LOOP_DETECTED,
+                                TS_HTTP_STATUS_NOT_EXTENDED,
+                                TS_HTTP_STATUS_NETWORK_AUTHENTICATION_REQUIRED};
+
+static bool should_log(TSHttpTxn txnp);
+static void log_request_line(TSMBuffer bufp, TSMLoc loc, const char *output_header);
+static void log_response_status_line(TSMBuffer bufp, TSMLoc loc, const char *output_header);
+static void log_headers(TSMBuffer bufp, TSMLoc loc, const char *output_header);
+static void log_full_transaction(TSHttpTxn txnp);
+static int log_requests_plugin(TSCont contp ATS_UNUSED, TSEvent event, void *edata);
+
+static bool
+should_log(TSHttpTxn txnp)
+{
+  TSMBuffer txn_resp_bufp;
+  TSMLoc txn_resp_loc;
+  TSHttpStatus resp_status;
+
+  if (TSHttpTxnClientRespGet(txnp, &txn_resp_bufp, &txn_resp_loc) != TS_SUCCESS) {
+    TSError("[log_requests] Couldn't retrieve server response header.");
+    return false;
+  }
+
+  // if resp_status is blacklisted, don't log
+  resp_status = TSHttpHdrStatusGet(txn_resp_bufp, txn_resp_loc);
+  for (int i = 0; i < blacklist_size; ++i) {
+    if (resp_status == blacklist[i])
+      return false;
+  }
+
+  // else, if resp_status is in errors, log it
+  for (int i = 0; i < (int)(sizeof(errors) / sizeof(TSHttpStatus)); ++i) {
+    if (resp_status == errors[i]) {
+      return true;
+    }
+  }
+
+  // fall through
+  return false;
+}
+
+const char *
+convert_http_version(const int version)
+{
+  char *ret;
+  if (version == TS_HTTP_VERSION(1, 0))
+    ret = "HTTP/1.0";
+  else if (version == TS_HTTP_VERSION(1, 1))
+    ret = "HTTP/1.1";
+  else if (version == TS_HTTP_VERSION(1, 2))
+    ret = "HTTP/1.2";
+  else if (version == TS_HTTP_VERSION(2, 0))
+    ret = "HTTP/2.0";
+  else {
+    ret = "(Unknown HTTP version)";
+  }
+
+  return ret;
+}
+
+static void
+log_request_line(TSMBuffer bufp, TSMLoc loc, const char *output_header)
+{
+  int method_len;
+  int url_len;
+  const char *version;
+  TSMLoc url_loc;
+
+  // parse method
+  const char *method = TSHttpHdrMethodGet(bufp, loc, &method_len);
+
+  // parse version
+  version = convert_http_version(TSHttpHdrVersionGet(bufp, loc));
+
+  // parse request line URL
+  TSHttpHdrUrlGet(bufp, loc, &url_loc);
+  const char *url = TSUrlStringGet(bufp, url_loc, &url_len);
+
+  // null terminate these strings so we can printf them
```

Comment:
I think I see why you strdup these, you really don't have to. You should do a pattern like
```C
TSError([log_requests] %s request line='%.*s %.*s %s'", output_header, method_len, method, url_len, url, version);
```